### ISSUE TITLE
Reintroduced Java 6 Support Doesn't work

### ISSUE THREAD

**Issue by ghost** [antoineco/aco-oracle_java#21]
Hi,

I have one, possibly two issues with the module... 

We have some build agents which require Java 6, 7 and 8 to be installed.

Previously I've been using this module to install 7 and 8 and the Puppetlabs module to install Java 6. This was working fine until I updated the module for version 101.

Then I started having Puppet reporting:
`Notice: /Stage[main]/Oracle_java::Install::Rpm/Package[jdk]/ensure: current_value 2000:1.6.0_45-fcs, should be 2000:1.8.0_101-fcs (noop)`

on every run... Obviously it's finding the 1.6.0 installed by the other module.

As I noticed your change logs state that you'd added back support for 1.6.0 I decided to try and install all three Java versions just via your module.

Unfortunately this is failing with the message:
`Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, oracle_java module does not support Java SE version 6 (yet) at /etc/puppetlabs/code/environments/production/modules/oracle_java/manifests/installation.pp:152:7 at ...`

So it looks like whilst 6 is included in init.pp and javalist.pp it's not included in the installation.pp The case $maj_version only handles 7 and 8 still.

For info this is how I'm currently initiating the install:

``` puppet
  class { 'oracle_java':
    version         => '8u101',
    type            => 'jdk',
    add_alternative => true,
    before          => Java_ks['ca.file:truststore8'],
  } ->
  oracle_java::installation { ['7u79','6u45']:
      type   => 'jdk',
      before => Java_ks['ca.file:truststore7'],
  }
```

**Comment by ghost** [comment]
Ah... I've just noticed the Change Logs states: (main installation only) for Java 6...

So presumably I can't install using the above then? That's also likely why the module is now incorrectly detecting Java 6 rpms...

So `Notice: /Stage[main]/Oracle_java::Install::Rpm/Package[jdk]/ensure: current_value 2000:1.6.0_45-fcs, should be 2000:1.8.0_101-fcs (noop)`

on every run now sees both the Java 6 rpm and 8 rpm:

```
jdk-1.6.0_45-fcs.x86_64
jdk1.8.0_101-1.8.0_101-fcs.x86_64
jdk1.8.0_91-1.8.0_91-fcs.x86_64
```

when in fact `/usr/java/jdk1.7.0_79` is present and appears to have been installed just fine.

All installations specified like this:

``` puppet
} ->
oracle_java::installation { ['7u79','6u45']:
  type => 'jdk',
  before => Java_ks['ca.file:truststore7'],
}
```

are actually installed via a tar.gz download rather than rpm currently (even though the rpm's and checksums are listed in the manifest.

So in terms of installing all three and no seeing any errors I presume I'd unfortunately need:

``` puppet
class { 'oracle_java':
  version => '6u45',
  type => 'jdk',
  add_alternative => true,
} ->
oracle_java::installation { ['7u79','8u101']:
  type => 'jdk',
  add_alternative => true,
}
```

That way the latest Java 8 should still be the default Java?

**Comment by antoineco** [comment]
As you noticed the support for Java 6 was provided on a best-effort basis in this release, that's why it can only be installed in the main class. Either using the Rpm or tar.gz distribution, but still only using the main class. I could add it to the `installation` type if there is a use case for it of course.

The last example you posted is correct, and if your `java` alternative is set to 'auto' (you can check with `update-alternatives --display java | head -1`) it will always default to the last version, due to its higher priority.

The "conflict" you see is normal, the main installation will always replace the currently installed version if Rpm installation is selected, because the OS won't accept two packages called `jdk`, so Puppet will either upgrade what already exists on the system or fail. Since you already have Java 6u45 installed, Puppet shouldn't do anything (see below)

## 

For testing purposes I applied your manifest in a blank environment (CentOS 7):

```
# puppet apply gh_issue21.pp 
Notice: Compiled catalog for cb39bbcef6db.local in environment production in 0.92 seconds
Notice: /Stage[main]/Oracle_java::Install::Rpm/Package[jdk]/ensure: created
Notice: /Stage[main]/Oracle_java::Alternative::Rpm/Exec[add java alternative 6u45]/returns: executed successfully
Notice: /Stage[main]/Main/Oracle_java::Installation[7u79]/Archive[/usr/java/jdk-7u79-linux-x64.tar.gz]/ensure: download archive from http://download.oracle.com/otn-pub/java/jdk/7u79-b15/jdk-7u79-linux-x64.tar.gz to /usr/java/jdk-7u79-linux-x64.tar.gz and extracted in /usr/java with cleanup
Notice: /Stage[main]/Main/Oracle_java::Installation[7u79]/Exec[add java alternative 7u79]/returns: executed successfully
Notice: /Stage[main]/Main/Oracle_java::Installation[8u101]/Archive[/usr/java/jdk-8u101-linux-x64.tar.gz]/ensure: download archive from http://download.oracle.com/otn-pub/java/jdk/8u101-b13/jdk-8u101-linux-x64.tar.gz to /usr/java/jdk-8u101-linux-x64.tar.gz and extracted in /usr/java with cleanup
Notice: /Stage[main]/Main/Oracle_java::Installation[8u101]/Exec[add java alternative 8u101]/returns: executed successfully
Notice: Applied catalog in 150.24 seconds
```

```
# java -version
java version "1.8.0_101"
Java(TM) SE Runtime Environment (build 1.8.0_101-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.101-b13, mixed mode)
```

```
# update-alternatives --display java
java - status is auto.
 link currently points to /usr/java/jdk1.8.0_101/bin/java

/usr/java/jdk1.6.0_45/bin/java - priority 1600045
 slave appletviewer: /usr/java/jdk1.6.0_45/bin/appletviewer
[...]

/usr/java/jdk1.7.0_79/bin/java - priority 1700079
 slave appletviewer: /usr/java/jdk1.7.0_79/bin/appletviewer
[...]

/usr/java/jdk1.8.0_101/bin/java - priority 1800101
 slave appletviewer: /usr/java/jdk1.8.0_101/bin/appletviewer
[...]

Current `best' version is /usr/java/jdk1.8.0_101/bin/java.
```

Other java versions are available in their respective location:

```
# ls -l /usr/java/

lrwxrwxrwx 1 root root       22 Jul 22 11:59 latest -> /usr/java/jdk1.6.0_45
lrwxrwxrwx 1 root root       16 Jul 22 11:59 default -> /usr/java/latest
drwxr-xr-x 7 root root     4096 Jul 22 12:20 jdk1.6.0_45
drwxr-xr-x 8   10  143     4096 Apr 10  2015 jdk1.7.0_79
drwxr-xr-x 8   10  143     4096 Jun 22 10:13 jdk1.8.0_101
[...]
```

```
# /usr/java/jdk1.7.0_79/bin/java  -version
java version "1.7.0_79"
Java(TM) SE Runtime Environment (build 1.7.0_79-b15)
Java HotSpot(TM) 64-Bit Server VM (build 24.79-b02, mixed mode)

# /usr/java/jdk1.6.0_45/bin/java -version
java version "1.6.0_45"
Java(TM) SE Runtime Environment (build 1.6.0_45-b06)
Java HotSpot(TM) 64-Bit Server VM (build 20.45-b01, mixed mode)
```

**Comment by ghost** [comment]
This is a curious limitation for the modules rpm version given Oracle keeps the RPM versioning for the packages very specific.

E.g. jdk1.8.0_101 is the rpm name and not jdk. Oracle's packaging means you can't yum update between Oracle Java minor versions because the packages include the minor version in the name.

So on my server for example:
rpm -qa | grep jdk
jdk-1.6.0_45-fcs.x86_64
jdk1.8.0_101-1.8.0_101-fcs.x86_64
jdk1.8.0_91-1.8.0_91-fcs.x86_64

So if I rpm query a specific rpm:
rpm -q jdk-1.6.0_45
jdk-1.6.0_45-fcs.x86_64

rpm -q jdk1.8.0_101
jdk1.8.0_101

So if $oracle_java::packagename: is set to match one of the above then it should be able to correctly match the specific version?

Either way perhaps our option will be to use format => tar.gz for all installs.

Our use case for all three versions might well be used elsewhere as well. As a company with a lot of Java based applications and development we have to support our old versions as well as new. So our build agents all need JDK6, 7 and 8 installed. 

I like this module as it handles the downloads of each minor release. It's also great that the alternatives are updated correctly. This is something the rpm's don't do a great job of.

If there is a way to make the module work with rpm package names correctly then that would be my preference. It's always nice being able to rpm -qa | grep jdk to quickly see all installed versions.

**Comment by antoineco** [comment]
Oracle changed the rpm naming in the Java 8u20 release, the package name is actually defined in the [init class](https://github.com/antoineco/aco-oracle_java/blob/master/manifests/init.pp#L137-L142). The problem arises for all versions **before** 8u20, where the package name is just `jre`/`jdk`, making it impossible to collocate multiple Java installations on the same machine. Therefore extra installations are set to use tar.gz archives in any circumstances, as explained in the [README](https://github.com/antoineco/aco-oracle_java#limitations).

Adding support for Java 6 as extra installation is on the road map for 2.7.1 then ;)

**Comment by ghost** [comment]
Thanks for the explanation. 

I'll look forward to 2.7.1. It's much appreciated!

**Comment by antoineco** [comment]
Fixed in [v2.7.1](https://github.com/antoineco/aco-oracle_java/releases/tag/2.7.1)

**Comment by ghost** [comment]
Nice one! Thanks :) I've already upgraded and confirmed it works perfectly.
